### ISSUE TITLE
Add allowThis to rxjs-no-unsafe-scope rule

### ISSUE THREAD

**Issue by kctang** [cartant/rxjs-tslint-rules#52]
I am using observables in methods within classes. Some of the functions do things like:

```
of(1, 2, 3).pipe(
  concatMap(val => {
    return this.doSomething(val)
  })
)
```

With `rxjs-no-unsafe-scope`, it says "Unsafe scopes are forbidden".

I think it is due to `this.doSomething()`. Would be great if we can `allowThis` (and have it enabled by default)...just like `allowParameters`.

**Comment by cartant** [comment]
I'll have to look into this - if you pardon the pun.

I have vague memories about intending to have the rule allow for method calls via `this`, but not property reads and writes.

I'm curious though, would your `doSomething` method actually use the `this` context in its implementation?

**Comment by cartant** [comment]
Are you absolutely sure that it's failing for method calls via `this`? I've added an expectation to a fixture and `this.foo()` doe not trigger the rule. What versions of `rxjs-tslint-rules`, RxJS and TypeScript are you using?

**Comment by kctang** [comment]
hmm... I will need time to double confirm this. Middle of another project now. Will update here once I re-check (in a few days).

Thanks for the quick response.

**Comment by cartant** [comment]
No worries. In the interim, I will update the fixtures to include non-failing expectations that demonstrate/document the rule's behaviour with method calls.

**Comment by cartant** [comment]
See https://github.com/cartant/rxjs-tslint-rules/commit/6d3aa9c5465bc5c11314647ce1ab34b39b279394 for the added expectations. Note that the method calls don't effect failures.

Once we've figured out whether or not this is the behaviour you are seeing, I'll happily add `allowProperties` (with a default of `false`) and `allowMethods` (with a default of `true`).

**Comment by kctang** [comment]
I did additional tests. With this code:

```
export class Demo {
  val: number

  doOkay_ReadWriteInTap () {
    this.val = 10
    return of(123, 456).pipe(tap(val => {
      console.log(this.val)
      this.val = val
    }))
  }

  doOkay_ReadInMap () {
    this.val = 10
    return of(123, 456).pipe(map(val => {
      console.log(this.val)
    }))
  }

  doNotOkay_WriteInMap () {
    this.val = 10
    return of(123, 456).pipe(map(val => {
      // next line fails linting
      this.val = val
    }))
  }

  // is similar to doNotOkay_WriteInMap() but why no linting error?
  doOkay_WriteInMapViaMethod () {
    this.val = 10
    return of(123, 456).pipe(map(val => {
      this.writeWithMethod(val)
    }))
  }

  writeWithMethod (val: number) {
    this.val = val
  }
}
```

This is my findings:

- The ONLY thing that cause `Unsafe scopes are forbidden` is `this.val = val` within the `map()` operator.
- Linting passed when it is enclosed within `tap()`.
- Linting also passed when I used the `doOkay_WriteInMapViaMethod()` style which is a bit odd (feels like a workaround for `doNotOkay_WriteInMap()`).

With these findings, I am actually not sure if `doNotOkay_WriteInMap()` should or should not cause linting error. I'm a bit confused on what 'should be the best practice' now... hehe...

**Comment by cartant** [comment]
The `allowDo` and `allowTap` options default to `true`, so, by default, all bets are off within those operators, as they are usually only used for side effects anyway. To have the rule applied within a `do` or a `tap`, the relevant option would have to be set to `false`.

The rule allows method calls, but I think an `allowMethods` option should be added so that they, too, can be flagged.

I'm surprised that read-only use in `doOkay_ReadInMap ` doesn't effect a failure. I'd say that's a bug.

The primary motivator for the rule was to flag the use of variables at outer scopes, but usage of `this` - as in `doNotOkay_WriteInMap` - was also something I wanted to flag, as it's a code smell, IMO.

The reason for allowing method calls was for things like AngularFire - where you might call a  function that returns a database observable to which you want to switch or whatever. Or a HTTP service with which you want to call `get`.

I think having `allowProperties` default to `false` and `allowMethods` default to `true` is reasonable. Then the user can change the behaviour if they want.

**Comment by kctang** [comment]
Thanks for the clarifications on `tap()` and all the above. Much clearer now.

Agree that `allowMethods` should default to true for the reason you described. In those scenarios, would `doOkay_WriteInMapViaMethod()` pass the rule (`writeWithMethod()` access class properties)?

Look forward to having `allowMethods` and `allowProperties` options. 👍

**Comment by cartant** [comment]
4.7.0 has been published and it includes the `allowMethods` and `allowProperties` options.

**Comment by kctang** [comment]
Tried using 4.7.0. Works great.

Just sharing one minor odd observation (in my code/usage). Since I am linting test code using the same config, the `unsafe scopes are forbidden` is showing up in unit test codes and I don't think I should fix those.

I decided to `/* tslint:disable:rxjs-no-unsafe-scope */` test files with this warning because I am basically using the variables/services created during setup/teardown of the unit tests. Seems like a fair compromise.

Cheers!
